File: algorithms/algorithms-cplusplus/leetcode/Question_0002_Add_Two_Numbers.cc

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    if (!l1 || !l2) {
      return l1 ? l1 : l2;
    }
    ListNode dummy(0);
    ListNode *p = &dummy;
    int carry = 0;
    while (l1 || l2 || carry) {
      carry += l1 ? l1->val : 0;
      carry += l2 ? l2->val : 0;
      p->next = new ListNode(carry % 10);
      p = p->next;
      carry /= 10;
      l1 = l1 ? l1->next : l1;
      l2 = l2 ? l2->next : l2;
    }
    return dummy.next;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0002_Add_Two_Numbers.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0002_Add_Two_Numbers {
    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null || l2 == null) {
            return l1 == null ? l2 : l1;
        }
        ListNode dummy = new ListNode(0);
        ListNode p = dummy;
        int carry = 0;
        while (l1 != null || l2 != null || carry > 0) {
            carry += l1 == null ? 0 : l1.val;
            carry += l2 == null ? 0 : l2.val;
            p.next = new ListNode(carry % 10);
            p = p.next;
            carry /= 10;
            l1 = l1 == null ? l1 : l1.next;
            l2 = l2 == null ? l2 : l2.next;
        }
        return dummy.next;
    }
}
